cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/customaction/BasicCustomActionContext.java

```diff
@@ -0,0 +1,155 @@
+ */
+public class BasicCustomActionContext extends AbstractContext implements CustomActionContext {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BasicCustomActionContext.class);
+  private final CustomActionSpecification customActionSpecification;
+  private final WorkflowProgramInfo workflowProgramInfo;
+  private final Metrics userMetrics;
+  private ProgramState state;
+
+  public BasicCustomActionContext(Program workflow, RunId runId, Arguments runtimeArguments,
+                                  CustomActionSpecification customActionSpecification,
+                                  WorkflowProgramInfo workflowProgramInfo,
+                                  MetricsCollectionService metricsCollectionService,
+                                  DatasetFramework datasetFramework, TransactionSystemClient txClient,
+                                  DiscoveryServiceClient discoveryServiceClient,
+                                  @Nullable PluginInstantiator pluginInstantiator) {
+
+    super(workflow, runId, runtimeArguments, customActionSpecification.getDatasets(),
+          getMetricCollector(workflow, runId.getId(), metricsCollectionService), datasetFramework, txClient,
+          discoveryServiceClient, false, pluginInstantiator);
+    this.customActionSpecification = customActionSpecification;
+    this.workflowProgramInfo = workflowProgramInfo;
+    if (metricsCollectionService != null) {
+      this.userMetrics = new ProgramUserMetrics(getProgramMetrics());
+    } else {
+      this.userMetrics = null;
+    }
+  }
+
+  @Nullable
+  private static MetricsContext getMetricCollector(Program program, String runId,
+                                                   @Nullable MetricsCollectionService service) {
+    if (service == null) {
+      return null;
+    }
+    Map<String, String> tags = Maps.newHashMap();
+    tags.put(Constants.Metrics.Tag.NAMESPACE, program.getNamespaceId());
+    tags.put(Constants.Metrics.Tag.APP, program.getApplicationId());
+    tags.put(ProgramTypeMetricTag.getTagName(program.getType()), program.getName());
+    tags.put(Constants.Metrics.Tag.WORKFLOW_RUN_ID, runId);
+    return service.getContext(tags);
+  }
+
+  @Override
+  public Metrics getMetrics() {
+    return userMetrics;
+  }
+
+  @Override
+  public Map<String, Plugin> getPlugins() {
+    return null;
+  }
+
+  @Override
+  public CustomActionSpecification getSpecification() {
+    return customActionSpecification;
+  }
+
+  /**
+   * Sets the current state of the program.
+   */
+  public void setState(ProgramState state) {
+    this.state = state;
+  }
+
+  @Override
+  public ProgramState getState() {
+    return state;
+  }
+
+  @Override
+  public void execute(TxRunnable runnable) throws TransactionFailureException {
+    final TransactionContext context = datasetCache.newTransactionContext();
+    try {
+      context.start();
+      runnable.run(datasetCache);
+      context.finish();
```

Comment:
does this need to be run in a finally? or is it handled by the abortTransaction calls?